internal/controllers/app_controller.go

```diff
@@ -270,12 +314,298 @@ func (r *AppReconciler) reconcile(ctx context.Context, app *ketchv1.App, logger
 		}
 	}
 
+	if len(app.Spec.Deployments) > 0 && !app.Spec.Canary.Active {
+		// use latest deployment and watch events for each process
+		latestDeployment := app.Spec.Deployments[len(app.Spec.Deployments)-1]
+		for _, process := range latestDeployment.Processes {
+			var dep appsv1.Deployment
+			if err := r.Get(ctx, client.ObjectKey{
+				Namespace: framework.Spec.NamespaceName,
+				Name:      fmt.Sprintf("%s-%s-%d", app.GetName(), process.Name, latestDeployment.Version),
+			}, &dep); err != nil {
+				return reconcileResult{
+					status:  v1.ConditionFalse,
+					message: fmt.Sprintf("failed to get deployment: %v", err),
+				}
+			}
+			err = r.watchDeployEvents(ctx, app, framework.Spec.NamespaceName, &dep, &process, r.Recorder)
+			if err != nil {
+				return reconcileResult{
+					status:  v1.ConditionFalse,
+					message: fmt.Sprintf("failed to get deploy events: %v", err),
+				}
+			}
+		}
+		// We useTimeout here to set reconcile.ReququeAfter in the Reconciler
+		// in order to ensure events actually get sent. It seems the lazyRecorder we use
+		// can stop with unhandled messages if the reconciler rapidly requeues.
+		return reconcileResult{
+			framework:  ref,
+			status:     v1.ConditionTrue,
+			useTimeout: true,
+		}
+	}
+
 	return reconcileResult{
 		framework: ref,
 		status:    v1.ConditionTrue,
 	}
 }
 
+// watchDeployEvents watches a namespace for events and, after a deployment has started updating, records events
+// with updated deployment status and/or healthcheck and timeout failures
+func (r *AppReconciler) watchDeployEvents(ctx context.Context, app *ketchv1.App, namespace string, dep *appsv1.Deployment, process *ketchv1.ProcessSpec, recorder record.EventRecorder) error {
+	cli, err := kubernetes.NewForConfig(r.Config)
+	if err != nil {
+		return err
+	}
+
+	opts := metav1.ListOptions{
+		FieldSelector: "involvedObject.kind=Pod",
+		Watch:         true,
+	}
+	watcher, err := cli.CoreV1().Events(namespace).Watch(ctx, opts) // requires "watch" permission on events in clusterrole
+	if err != nil {
+		return err
+	}
+	defer watcher.Stop()
+
+	// wait for Deployment Generation
+	timeout := time.After(DefaultPodRunningTimeout)
+	for dep.Status.ObservedGeneration < dep.Generation {
+		dep, err = cli.AppsV1().Deployments(namespace).Get(ctx, dep.Name, metav1.GetOptions{})
+		if err != nil {
+			recorder.Eventf(app, v1.EventTypeWarning, appReconcileError, "error getting deployments: %s", err.Error())
+			return err
+		}
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-timeout:
+			recorder.Event(app, v1.EventTypeWarning, appReconcileError, "timeout waiting for deployment generation to update")
+			return errors.Errorf("timeout waiting for deployment generation to update")
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	go r.watchFunc(ctx, app, namespace, dep, process, recorder, watcher, cli, timeout)
```

Comment:
do processes share watcher?

we instantiate watcher for each process, but i would say from code that we would get same events for each process, we just print them differently `Updating units [%s]", process.Name`

how do we know that some events belong to certain process?

Reply:
Each process has it's own watcher (is that a good idea 🤷 ). The watcher does 2 things: 1) watch Pod Events and 2) keep checking the appsv1.Deployment for updates. 

1. The Pod Events are filtered by process [here](https://github.com/theketchio/ketch/pull/177/files#diff-01570fd749623ed07d5f2e0b7097495a4ef86b6f3419378180bc43fc73c9223eR547) because the appsv1.Deployment name includes the process name. An appsv1.Deployment is 1-to-1 with a Ketch Process (which confuses because Ketch has Deployments too). 
2. The appsv1.Deployments that we keep checking are also filtered by appsv1.Deployment.Name [e.g. here](https://github.com/theketchio/ketch/pull/177/files#diff-01570fd749623ed07d5f2e0b7097495a4ef86b6f3419378180bc43fc73c9223eR481) which is the Ketch Process Name.

At least, I think that's what's going on. I'm new here.